File: examples/python/optimize_custom_type.py
"""
An example of defining a vector space manifold R^2
In Python this is done by defining manifold-related member functions,
  1. dim() function returns manifold dimensionality, 
  2. local() and retract() functions defines the local coordinate chart.
  3. __repr__() for printing
"""

from __future__ import print_function

from minisam import *
import numpy as np


# ############################## custom type ###################################

class Point2D(object):
    # constructor
    def __init__(self, x, y):
        self.x_ = float(x)
        self.y_ = float(y)

    # print function
    def __repr__(self):
        return 'custom 2D point [' + str(self.x_) + ', ' + str(self.y_) + ']\''

    # local coordinate dimension
    def dim(self):
        return 2

    # map manifold point other to local coordinate
    def local(self, other):
        return np.array([other.x_ - self.x_, other.y_ - self.y_], dtype=np.float)

    # apply changes in local coordinate to manifold, \oplus operator
    def retract(self, vec):
        return Point2D(self.x_ + vec[0], self.y_ + vec[1])


# ################### optimizing custom 2D point type ##########################

# graph container
graph = FactorGraph()

# add a single prior on (0, 0)
graph.add(PriorFactor(key('x', 0), Point2D(0, 0), None))

print(graph, '\n')


# initial variables for optimization
initials = Variables()

# initial point value set to (2, 3)
initials.add(key('x', 0), Point2D(2, 3))

print("initial:\n", initials, '\n')


# optimize
opt_param = GaussNewtonOptimizerParams()
opt = GaussNewtonOptimizer(opt_param)

results = Variables()

status = opt.optimize(graph, initials, results)
if status != NonlinearOptimizationStatus.SUCCESS:
    print("optimization error : ", status)

print("optimized:\n", results, '\n')

File: examples/cpp/optimize_custom_type.cpp
/**
 * An example of defining a vector space manifold R^2
 * In C++ this is done by specializing minisam::traits<> with
 *   1. minisam type category tag
 *   2. tangent space vector type defs
 *   3. Dim() function returns manifold dimensionality,
 *   4. Local() and Retract() functions defines the local coordinate chart.
 *   5. Print() for printing
 */

#include <minisam/core/FactorGraph.h>
#include <minisam/core/Variables.h>
#include <minisam/nonlinear/GaussNewtonOptimizer.h>
#include <minisam/slam/PriorFactor.h>

#include <iostream>

/* ************************* example custom type **************************** */

class Point2D {
 private:
  double x_, y_;

 public:
  // constructor
  Point2D(double x, double y) : x_(x), y_(y) {}

  // access
  double x() const { return x_; }
  double y() const { return y_; }
};

/* ************************** custom type traits **************************** */

namespace minisam {
template <>
struct traits<Point2D> {
  // minisam type category tag, used for static checking
  // must be manifold_tag or lie_group_tag to be optimizable
  typedef manifold_tag type_category;

  // print
  static void Print(const Point2D& m, std::ostream& out = std::cout) {
    out << "custom 2D point [" << m.x() << ", " << m.y() << "]'";
  }

  // tangent vector type defs
  typedef Eigen::Matrix<double, 2, 1> TangentVector;

  // local coordinate dimension
  static constexpr size_t Dim(const Point2D&) { return 2; }

  // map manifold point s to local coordinate
  static TangentVector Local(const Point2D& origin, const Point2D& s) {
    return Eigen::Matrix<double, 2, 1>(s.x() - origin.x(), s.y() - origin.y());
  }

  // apply changes in local coordinate to manifold, \oplus operator
  static Point2D Retract(const Point2D& origin, const TangentVector& v) {
    return Point2D(origin.x() + v[0], origin.y() + v[1]);
  }
};
}  // namespace minisam

/* ******************* optimizing custom 2D point type ********************** */

using namespace std;
using namespace minisam;

int main() {
  // graph container
  FactorGraph graph;

  // add a single prior on (0, 0)
  graph.add(PriorFactor<Point2D>(key('x', 0), Point2D(0, 0), nullptr));

  graph.print();
  cout << endl;

  // initial variables for optimization
  Variables initials;

  // initial point value set to (2, 3)
  initials.add(key('x', 0), Point2D(2, 3));

  cout << "initials:" << endl;
  initials.print();
  cout << endl;

  // optimize
  GaussNewtonOptimizer opt(GaussNewtonOptimizerParams{});
  Variables results;

  auto status = opt.optimize(graph, initials, results);
  if (status != NonlinearOptimizationStatus::SUCCESS) {
    cout << "optimization error" << endl;
  }

  cout << "optimized:" << endl;
  results.print();
  cout << endl;

  return 0;
}
